### PR TITLE
Create TypedNeoDateTimeFormatter and DateTimePattern types

### PR DESCRIPTION
**Pull request by sffc**
Part of #3347
Depends on #4496
Depends on #4497

### CODE REVIEW

File: components/datetime/src/raw/neo.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    #[test]
+    fn test_sizes() {
+        assert_eq!(48, core::mem::size_of::<DataPayload<ErasedDatePatternV1Marker>>());
+        assert_eq!(80, core::mem::size_of::<DataLocale>());
+        assert_eq!(216, core::mem::size_of::<FixedDecimalFormatter>());
+        assert_eq!(2, core::mem::size_of::<WeekCalculator>());
+        assert_eq!(72, core::mem::size_of::<DataPayload<ErasedYearNamesV1Marker>>());
+        assert_eq!(56, core::mem::size_of::<DataPayload<ErasedMonthNamesV1Marker>>());
+        assert_eq!(40, core::mem::size_of::<DataPayload<WeekdayNamesV1Marker>>());
+        assert_eq!(40, core::mem::size_of::<DataPayload<DayPeriodNamesV1Marker>>());
```
Comment (sffc) [inline components/datetime/src/raw/neo.rs:65]: Thought: We could save space in DateTimeNames by making this `LiteMap<Field, DataNamesPayload, ShortVec>` where `DataNamesPayload` is an enum between the payload types. It compels an allocation however for a pattern containing more than 1 display name. @Manishearth

**Comment by sffc** [comment]
Background: the current DateTimeFormatter is over 6000 bytes. I've demonstrated in this PR that NeoDateTimeFormatter is around 700 bytes, which is a huge improvement, but we want to see if we can get it even smaller.

File: components/datetime/src/raw/neo.rs (hunk 1 above)
Comment (Manishearth) [reply components/datetime/src/raw/neo.rs:65]: Most patterns need more than one display name, right?

**Comment by sffc** [comment]
Now that I have an end-to-end formatter I started looking at performance numbers. Neo hits the data provider more than the old, since there are more payloads to load, but it reduces the amount of processing once the payloads are loaded. The pattern formatting itself appears to be mostly identical. The overall result is that Neo is a bit faster than Old for date-only formats (where less data needs to be loaded) and a bit slower than Old for larger datetime formats (where multiple payloads need to be loaded). In the fixture test, the overall Criterion performance shows about 5% slower for Neo, with the hottest function being `DataLocale::strict_cmp`, which is quite fast by itself (it's something I've already hyper-optimized) but it gets called 1712 times in my little test code, which creates 32 formatters in a variety of locales and lengths. `DataLocale::clone` shows up too but it is tiny compared to the comparison function. `LocaleFallbackIterator::step` also shows up and is fairly heavy but it isn't called as many times as `strict_cmp`.

Side note: I'm trying to figure out why `strict_cmp` gets called so many times. I see it called in the following places: (1) `get_explicit_parent`: while binary searching over the sorted list of parents, which happens at each step, and (2) Bake provider during the lookup function. Note that blob provider no longer uses `strict_cmp` since the migration to ZeroTrie. If we also migrate the bake provider to ZeroTrie, it may help here.

In other words, it seems any bottlenecks are outside of the datetime code; within the datetime code, Neo is equal to or outperforms Old. So I would like to proceed.

File: components/datetime/src/raw/neo.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/src/raw/neo.rs]: Well patterns containing only AM/PM ("9:46 pm" or "12/26/2023 at 9:46 pm") or only the month name ("Dec 26, 2023") are fairly common, but yes for anything more than that we would allocate which isn't great. But, saving stack space is nice. I don't want to do this now, but something to at least consider in #4337

**Comment by sffc** [comment]
Review tips:

Most of this PR is plumbing. I tried to pull off the interesting parts as I was writing it. There are two more PRs, #4496 and #4497, that are currently open with two small but slightly more interesting bits of code.

This PR adds the following public types:

1. `icu_datetime::neo_pattern::DateTimePattern`, a public wrapper around `icu_datetime::pattern::runtime::Pattern`, with a more restricted API surface sufficient to be used in `icu_datetime::TypedDateTimeNames` (which is already landed)
2. `icu_datetime::neo::TypedNeoDateFormatter`, which supports formatting with date lengths and uses all neo data
3. `icu_datetime::neo::TypedNeoTimeFormatter`, which supports formatting with time lengths and uses all neo data
4. `icu_datetime::neo::TypedNeoDateTimeFormatter`, which supports formatting with date and time lengths and uses all neo data

The code in `src/neo.rs` and `src/neo_pattern.rs` is intended to be as thin as reasonably possible, with all business logic living in other modules such as the new `src/raw/neo.rs`. The code in `src/neo.rs` is mostly data provider marshaling; by the time you get into `src/raw/neo.rs` the data provider types are normalized for you.

File: components/datetime/src/external_loaders.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+use icu_calendar::CalendarError;
+use icu_decimal::options::FixedDecimalFormatterOptions;
+use icu_decimal::provider::DecimalSymbolsV1Marker;
+use icu_decimal::{DecimalError, FixedDecimalFormatter};
+use icu_provider::prelude::*;
+
+/// Trait for loading a FixedDecimalFormatter.
+///
+/// Implemented on the provider-specific loader types in this module.
+pub(crate) trait FixedDecimalFormatterLoader {
```
Comment (Manishearth) [inline components/datetime/src/external_loaders.rs:18]: observation: must stay private, we may need to be able to load RBNF in the long run

in the medium term we will probably hardcode a `FixedDecimalFormatter | CustomFormat` where the CustomFormat is a simple enum with code attached.

File: components/datetime/src/neo_pattern.rs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+///
+/// [`DateTimeFormatter`]: crate::DateTimeFormatter
+/// [`FormattedNeoDateTime::pattern`]: crate::neo::FormattedNeoDateTime::pattern
+#[derive(Debug)]
+pub struct DateTimePattern {
+    pattern: runtime::Pattern<'static>,
+}
+
+impl DateTimePattern {
+    /// Creates a [`DateTimePattern`] from a pattern string.
```
Comment (Manishearth) [inline components/datetime/src/neo_pattern.rs:58]: nit (nb, followup): roughly document thr format or link to UTS 35.

**Review by Manishearth — DISMISSED** [review]

File: components/datetime/src/format/neo.rs [hunk 4]
```diff
@@ -855,53 +874,78 @@ impl RawDateTimeNames {
...
-            -> Result<WeekCalculator, icu_calendar::CalendarError>,
-    ) -> Result<RawDateTimePatternFormatter<'l>, Error>
+    /// Loads all data required for formatting the given [`PatternItem`]s.
+    ///
+    /// This function has a lot of arguments because many of the arguments are generic,
+    /// and pulling them out to an options struct would be cumbersome.
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn load_for_pattern<YearMarker, MonthMarker>(
+        &mut self,
+        year_provider: Option<&(impl DataProvider<YearMarker> + ?Sized)>,
```
[the remark this answers is not in the document]
Comment (sffc) [reply components/datetime/src/format/neo.rs:909]: Hmm, with AnyCalendar `<YearMarker, MonthMarker>` are not known until runtime and I don't really want to include more than 1 copy of this function (i.e. through an enum switch). Maybe we can move those over to the external loader architecture so that the enum switch is isolated to the load function. Let's figure that out when I implement the AnyCalendar version of neo formatter.

[the remark this answers is not in the document]
Comment (Manishearth) [reply components/datetime/src/format/neo.rs:909]: Yep.

**Review by Manishearth — APPROVED** [review]
